jsbits/diff.js

```diff
@@ -198,168 +205,152 @@ function syncChildren (os, ns, parent) {
+            while (os[oldFirstIndex].key === ns[newFirstIndex].key && os[oldLastIndex].key === ns[newLastIndex].key) {
+              diff (os[oldLastIndex], ns[newLastIndex], parent);
+	      diff (os[oldFirstIndex], ns[newFirstIndex], parent);
+	      oldLastIndex--; newLastIndex--;
+ 	      oldFirstIndex++; newFirstIndex++;
+	      if (oldFirstIndex > oldLastIndex || newFirstIndex > newLastIndex) break;
+	    }
+
+	    while (os[oldFirstIndex].key === ns[newFirstIndex].key) {
+	      diff (os[oldFirstIndex], ns[newFirstIndex], parent);
+ 	      oldFirstIndex++; newFirstIndex++;
+	      if (oldFirstIndex > oldLastIndex || newFirstIndex > newLastIndex) break;
+	    }
+
+            while (os[oldLastIndex].key === ns[newLastIndex].key) {
+              diff (os[oldLastIndex], ns[newLastIndex], parent);
+	      oldLastIndex--; newLastIndex--;
+	      if (oldFirstIndex > oldLastIndex || newFirstIndex > newLastIndex) break;
+	    }
+        }
 	/* flip-flop case, nodes have been swapped, in some way or another
 	   both could have been swapped.
 	   -> [ a b c ] <- old children
-	   -> [ c b a ] <- new children */
+	   -> [ c b a ] <- new children
+	*/
 	else if (oFirst.key === nLast.key && nFirst.key === oLast.key) {
-	    var nextSib = oFirst.domRef.nextSibling;
-	    parent.insertBefore(oFirst.domRef, oLast.domRef);
-	    parent.insertBefore(nextSib, oLast.domRef);
-	    newFirstIndex++;
-	    oldFirstIndex++;
-	    oldLastIndex--;
-	    newLastIndex--;
-	    continue;
-	}
+	    // swap array elements
+	    tmp = oFirst;
+	    oFirst = oLast;
+	    oLast = tmp;
 
-	/* or just one could be swapped (d's align here)
-	   -> [ d b g ] <- old children
-	   -> [ a k d ] <- new children
-	   on either side (e's align here)
-	   -> [ e b c ] <- old children
-	   -> [ b c e ] <- new children */
+	    // swap pointers
+	    node = oLast.domRef.nextSibling;
+	    parent.insertBefore(oLast.domRef, oFirst.domRef.nextSibling);
+	    parent.insertBefore(oFirst.domRef, node);
 
-	/* For now, the above case is handled in the "you're screwed case" below. */
+	    // diff
+            diff (oLast, nLast, parent);
+            diff (oFirst, nFirst, parent);
+
+	    // incr/decr
+	    newFirstIndex++; oldFirstIndex++;
+	    oldLastIndex--; newLastIndex--;
+	}
+	/* Or just one could be swapped (d's align here)
+           This is top left and bottom right match case.
+           We move d to end of list, mutate old vdom to reflect the change
+	   We then continue without affecting indexes, hoping to land in a better case
+	   -> [ d a b ] <- old children
+	   -> [ a b d ] <- new children
+	   becomes
+	   -> [ a b d ] <- old children
+	   -> [ a b d ] <- new children
+	   and now we happy path
+        */
+        else if (oFirst.key === nLast.key) {
+	    parent.insertBefore(oFirst.domRef, oLast.domRef.nextSibling);
+	    /* swap positions in old vdom */
+   	    os.splice(oldLastIndex, 0, os.splice(oldFirstIndex, 1)[0]);
```

Comment:
IMHO this should be wrapped up in some kind of `swap` function. I also suspect that just assigning things to a temporary variable and doing the swap that way is going to be faster (and for me personally it’s easier to read) but I haven’t measured this.

Reply:
agreed